File: python/backtracking/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = set()

        def backtracker(remaining, constructed):
            if len(remaining) == 0:
                answer.add(tuple(constructed[:]))
                return

            current = set(list(remaining)[:])

            for i in remaining:
                current.remove(i)
                backtracker(current, constructed + [i])
                current.add(i)

        remaining = set(nums)
        backtracker(remaining, [])
        finalanswer = []
        for i in answer:
            finalanswer.append(list(i))
        return finalanswer

File: c++/backtracking/permutations.cc
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        std::vector<std::vector<int>> permutations;
        construct(permutations, {}, nums);

        return permutations;
    }

private:
    void construct(std::vector<std::vector<int>>& permutations, std::vector<int> permutation, std::vector<int>& nums) {
        if (permutation.size() == nums.size()) {
            permutations.push_back(permutation);
            return;
        }

        for (int num : nums) {
            if (std::find(permutation.begin(), permutation.end(), num) == permutation.end()) {
                permutation.push_back(num);
                construct(permutations, permutation, nums);
                permutation.pop_back();
            }
        }
    }
};
